yb-voyager/test/containers/testcontainers.go

```diff
@@ -0,0 +1,142 @@
+package testcontainers
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/samber/lo"
+	log "github.com/sirupsen/logrus"
+)
+
+// TODO: Implement containerRegistry to have Singleton Pattern for the containers
+// containerRegistry to ensure one container per database(dbtype+version) [Singleton Pattern]
+var (
+	containerRegistry map[string]TestContainer
+	registryMutex     sync.Mutex
+)
+
+type TestContainer interface {
+	Start(ctx context.Context) error
+	Terminate(ctx context.Context)
+	GetHostPort() (string, int, error)
+	GetConfig() ContainerConfig
+	GetConnectionString() string
+	/*
+		TODOs
+			// Function to run sql script for a specific test case
+			SetupSqlScript(scriptName string, dbName string) error
+
+			// Add Capability to not detect and not spawn the cluster if already running
+			// Add Capability to run multiple versions of a dbtype parallely
+	*/
+}
+
+type ContainerConfig struct {
+	DBVersion string
+	User      string
+	Password  string
+	DBName    string
+	Schema    string
+}
+
+func NewTestContainer(dbType string, containerConfig *ContainerConfig) TestContainer {
+	registryMutex.Lock()
+	defer registryMutex.Unlock()
+
+	// initialise containerConfig struct if nothing is provided
+	if containerConfig == nil {
+		containerConfig = &ContainerConfig{}
+	}
+	setContainerConfigDefaultsIfNotProvided(dbType, containerConfig)
+
+	// check if container is already created after fetching default configs
+	containerName := fmt.Sprintf("%s-%s", dbType, containerConfig.DBVersion)
+	if container, exists := containerRegistry[containerName]; exists {
+		log.Infof("container '%s' already exists in the registry", containerName)
+		return container
+	}
+
+	var testContainer TestContainer
+	switch dbType {
+	case POSTGRESQL:
+		testContainer = &PostgresContainer{
+			ContainerConfig: *containerConfig,
+		}
+	case YUGABYTEDB:
+		testContainer = &YugabyteDBContainer{
+			ContainerConfig: *containerConfig,
+		}
+	case ORACLE:
+		testContainer = &OracleContainer{
+			ContainerConfig: *containerConfig,
+		}
+	case MYSQL:
+		testContainer = &MysqlContainer{
+			ContainerConfig: *containerConfig,
+		}
+	default:
+		panic(fmt.Sprintf("unsupported db type '%q' for creating test container\n", dbType))
+	}
+
+	return testContainer
```

Comment:
we should add to the containerRegistry here, before returning, right?

Reply:
oh right, my bad.